### PR TITLE
v4 React on Rails: redux_store from either view or controller, other fixes

### PR DESCRIPTION
**Pull request by justin808**
* Updated API of ReactOnRails.js to provide optional param for getStore.
* Added API docs for new controller module and view helper for shared
  redux stores.
* Deprecate usage of redux_store from a view helper.
* Adding other helpers and JS code extracted from the https://github.com/shakacode/react-webpack-rails-tutorial/

[Review on Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/311)

### CODE REVIEW

File: node_package/src/StoreRegistry.js [hunk 1]
```diff
@@ -24,16 +24,22 @@ export default {
...
-      console.log('storeKeys', storeKeys);
-      throw new Error(`Could not find hydrated store with name '${name}'. \
-Hydrated store names include [${storeKeys}].`);
+      if (throwIfMissing)  {
+        const storeKeys = Array.from(_stores.keys()).join(', ');
+        console.log('storeKeys', storeKeys);
+        throw new Error(`Could not find hydrated store with name '${name}'. ` +
+          `Hydrated store names include [${storeKeys}].`);
+      } else {
+        return;
```
Comment (alex35mil) [inline node_package/src/StoreRegistry.js:41]: Linter complains here.

File: node_package/tests/StoreRegistry.test.js [hunk 2]
```diff
@@ -1,8 +1,7 @@
-
 import test from 'tape';
 import StoreRegistry from '../src/StoreRegistry';
 import React from 'react';
-import { createStore } from 'redux';
+import {createStore} from 'redux';
```
Comment (alex35mil) [inline node_package/tests/StoreRegistry.test.js]: We usually use spaces here.

**Comment by alexfedoseev** [comment]
LGTM! We should try it out :+1:

**Comment by justin808** [comment]
CI is failing when I do a fresh git clone, but not in my regular development setup.

**Comment by justin808** [comment]
It may not be worth deprecating the view based setup, and if so, the version would go to 3.1 rather than 4.0.

**Comment by lucke84** [comment]
Took a quick look, maybe it's worth trying and upgrade node to 5.6.0?

**Comment by seoyoochan** [comment]
@justin808 if we only pass props as `redux_store("helloWorldStore", @hello_world_props)` in controller, without passing them to a react component as `<%= react_component("HelloWorldApp", props: @hello_world_props, prerender: true) %>`, you will have the following errors..

```
[SERVER] helloWorldStore.jsx {"name":"Stranger"}
[SERVER] RENDERED HelloWorldApp to dom node with id: HelloWorldApp-react-component-0
[SERVER] HelloWorldAppServer {}
[SERVER] helloWorldStore.jsx {}
[SERVER] Warning: Failed propType: Required prop `name` was not specified in `HelloWorldWidget`. Check the render method of `HelloWorld`.
helloWorldStore.jsx {"name":"Stranger"}
RENDERED HelloWorldApp to dom node with id: HelloWorldApp-react-component-0
HelloWorldAppClient {}
helloWorldStore.jsx {}
Warning: Failed propType: Required prop `name` was not specified in `HelloWorldWidget`. Check the render method of `HelloWorld`.
 ``` 
In both *HelloWorldAppServer.jsx* and *HelloWorldAppClient.jsx*,

`props` becomes empty object if the react component has not been given any props from its Rails view. 
if I pass props to react component as well, it works finally.

File: node_package/tests/StoreRegistry.test.js (hunk 2 above)
Comment (justin808) [inline node_package/tests/StoreRegistry.test.js]: Linter should be catching this.

**Comment by justin808** [comment]
I'm going to apply this to the https://github.com/shakacode/react-webpack-rails-tutorial and then merge if that goes well. NOTE: There are breaking changes.

Please try this out:

* https://rubygems.org/gems/react_on_rails/versions/4.0.0.beta.3
* https://www.npmjs.com/package/react-on-rails

**Pull request merged**